rust/generated_parser/src/ast_builder.rs

```diff
@@ -3789,4 +4047,366 @@ impl<'alloc> AstBuilder<'alloc> {
     ) -> Result<'alloc, arena::Box<'alloc, Void>> {
         Err(ParseError::NotImplemented("export"))
     }
+
+    // Note binding info to the stack.
+    fn on_binding_identifier(&mut self, name: &'alloc str, offset: usize) {
+        match self.bindings.last() {
+            Some(info) => {
+                debug_assert!(info.offset < offset);
+            }
+            None => {}
+        }
+
+        self.bindings.push(BindingInfo {
+            name,
+            offset,
+            kind: BindingKind::Unknown,
+        });
+    }
+
+    // Update binding kind in [from, to) range of source position.
```

Comment:
This needs a comment that explains why.

// Update the binding kind of all names declared in a specific range of the source
// (and not in any nested scope). This is used e.g. when the parser reaches the
// end of a VariableStatement to mark all the variables as Var bindings.
//
// It's necessary because the current parser only calls AstBuilder methods at
// the end of each production, not at the beginning.
//
// Every binding must be marked using this method before we reach the end
// of its scope.

Reply:
Or, just point to the comment at `AstBuilder::bindings` if you think that's better.

Reply:
actually, not all binding have to be marked.

for example, `CatchParameter` don't have to be marked, given `Catch` syntax knows the offset range.
same for `FormalParameter`.

Bindings that have to be marked are inside `StatementList`, given multiple binding kinds can appear there and the enclosing context doesn't know where they are,

Reply:
will add comments about that to `BindingKind`